tools/serverpod_cli/lib/src/generator/types.dart

```diff
@@ -418,20 +425,23 @@ class TypeDefinition {
   /// protocol: prefix in types. Whenever no url is set and user specified a
   /// class/enum with the same symbol name it defaults to the protocol: prefix.
   TypeDefinition applyProtocolReferences(
-      List<SerializableModelDefinition> classDefinitions) {
+    List<SerializableModelDefinition> classDefinitions,
+  ) {
+    var modelDefinition =
+        classDefinitions.where((c) => c.className == className).firstOrNull;
+    bool isProjectModel =
+        url == defaultModuleAlias || (url == null && modelDefinition != null);
     return TypeDefinition(
         className: className,
         nullable: nullable,
         customClass: customClass,
         dartType: dartType,
+        projectModelDefinition: isProjectModel ? modelDefinition : null,
```

Comment:
Question: Do you actually need this check? Wouldn't sending in the modelDefinition directly effectively have the same behaviour?